lib/ProductOpener/Ingredients.pm

```diff
@@ -2542,6 +2546,67 @@ sub init_percent_values ($total_min, $total_max, $ingredients_ref) {
 	return;
 }
 
+=head2 set_percent_max_from_taxonomy ( ingredients_ref )
+
+Set the percentage maximum for ingredients like flavouring where this is defined
+on the Ingredients taxonomy. The percent_max will not be applied in the following cases:
+
+ - if applying the percent_max would mean that it is not possible for the ingredient
+   total to add up to 100%
+ - If a later ingredient has a higher percentage than the percent_max of the restricted ingredient
+
+=cut
+
+sub set_percent_max_from_taxonomy ($ingredients_ref) {
+	# Exit if the first ingredient is constrained
+	if (!@{$ingredients_ref}
+		|| defined get_inherited_property("ingredients", $ingredients_ref->[0]{id}, "percent_max:en"))
+	{
+		return;
+	}
+
+	# Loop backwards through ingredients, checking that we don't set a percent_max that
+	# would be lower than the sum of the ingredients that come afterwards
+	my $current_max = 0;
+	for (my $index = scalar @{$ingredients_ref} - 1; $index > 0; $index--) {
+		my $ingredient = $ingredients_ref->[$index];
+		my $current_percent = $ingredient->{percent};
+		if (defined $current_percent) {
+			if ($current_percent > $current_max) {
+				$current_max = $current_percent;
+			}
+		}
+		else {
+			# See if taxonomy defines a maximum percent
+			my $percent_max = get_inherited_property("ingredients", $ingredient->{id}, "percent_max:en");
+			if (defined $percent_max and $percent_max > $current_max) {
+				# Maximum percantage for ingredients like flavourings
+				$ingredient->{percent_max} = $percent_max;
+				$current_max = $percent_max;
+			}
+		}
+	}
+
+	# Loop forwards through the ingredients to make sure that the maximum
+	# does not limit preceding ingredients where percent is specified
+	my $remaining_percent = 100;
+	for my $ingredient (@{$ingredients_ref}) {
+		my $defined_percent = $ingredient->{percent};
+		if (!defined $defined_percent) {
+			my $percent_max = $ingredient->{percent_max};
+			if (defined $percent_max && $percent_max < $remaining_percent) {
+				delete $ingredient->{percent_max};
```

Comment:
I think this can be removed (unless I'm mistaken), but if it is needed, then we could set percent_max to $remaining_percent instead of deleting it.